python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+            s = "[{}, {}]".format(repr(self.head), repr(self.tail))
+        else:
+            s = "[{}, ..., {}]".format(repr(self.head), repr(self.tail))
+        return s
+
+    def __len__(self):
+        # Returns the number of segments in the list.
+        count = 0
+        seg = self.head
+        while seg is not None:
+            count += 1
+            seg = seg.next
+        return count
+
+    def add(self, other):
+        """
+        Use to append another SegmentList, or a single segment.
+        SegmentList1.add(SegmentList2) will modify SegmentList1 so that
+        SegmentList1.tail.next = SegmentList2.head
+        SegmentList1.add(Segment1) will add Segment1 to the tail of SegmentList1
+        """
+        assert isinstance(other, SegmentList) or isinstance(other, Segment)
+
+        if isinstance(other, SegmentList):
+            if self.head is None:
+                self.head = other.head
+                self.tail = other.tail
+            else:
+                self.tail.next = other.head
+                self.tail = other.tail
+        elif isinstance(other, Segment):
+            if self.head is None:
+                self.head = other
+                self.tail = other
+            else:
+                self.tail.next = other
+                self.tail = other
+
+
+class IbdFinder:
+    """
+    Finds all IBD relationships between specified samples in a tree sequence.
+    """
+
+    def __init__(self, ts, samples=None, min_length=0, max_time=None):
+
+        self.ts = ts
+        # Note: samples *must* be in order of ascending node ID
+        if samples is None:
+            self.samples = ts.samples()
+        else:
+            self.samples = samples
+        self.samples.sort()
+        self.min_length = min_length
+        self.max_time = max_time
+        self.current_parent = self.ts.tables.edges.parent[0]
+        self.current_time = 0
+        self.A = [None for _ in range(ts.num_nodes)]  # Descendant segments
+        self.tables = self.ts.tables
+        self.parent_list = [[] for i in range(0, self.ts.num_nodes)]
+        for e in self.ts.tables.edges:
+            if (
+                len(self.parent_list[e.child]) == 0
+                or e.parent != self.parent_list[e.child][-1]
+            ):
+                self.parent_list[e.child].append(e.parent)
+        # Objects below are needed for the IBD segment-holding object.
+        self.sample_id_map = self.get_sample_id_map()
+        self.num_samples = len(self.samples)
+        self.sample_pairs = self.get_sample_pairs()
+        self.PARENT_SHOULD_BE_ADDED = True
+
+        # Note: in the C code the object below should be a struct array.
+        # Each item will be accessed using its index, which corresponds to a particular
+        # sample pair. The mapping between index and sample pair is defined in the
+        # find_sample_pair_index method further down.
+
+        self.ibd_segments = {}
+        for key in self.sample_pairs:
```

Comment:
I don't think there's much point in storing None for all the sample_pairs, right? We can just check to see if the key is already in the array or not.

Reply:
Let's talk about this